lxd/api_cluster.go

```diff
@@ -3011,207 +3081,109 @@ func evacuateClusterMember(d *Daemon, r *http.Request, mode string) response.Res
 			_ = evacuateClusterSetState(d, nodeName, db.ClusterMemberStateCreated)
 		})
 
-		metadata := make(map[string]any)
-
 		ctx := context.TODO()
 
-		for _, inst := range instances {
-			instProject := inst.Project()
-			l := logger.AddContext(logger.Log, logger.Ctx{"project": instProject.Name, "instance": inst.Name()})
-
-			// Check if migratable.
-			migrate, live := inst.CanMigrate()
-
-			// Apply overrides.
-			if mode != "" {
-				if mode == "stop" {
-					migrate = false
-					live = false
-				} else if mode == "migrate" {
-					migrate = true
-					live = false
-				} else if mode == "live-migrate" {
-					migrate = true
-					live = true
-				}
-			}
+		err = evacuateInstances(ctx, d, r, instances, mode, nodeName, stopInstance, migrateInstance, op)
+		if err != nil {
+			return err
+		}
 
-			// Stop the instance if needed.
-			isRunning := inst.IsRunning()
-			if isRunning && !(migrate && live) {
-				metadata["evacuation_progress"] = fmt.Sprintf("Stopping %q in project %q", inst.Name(), instProject.Name)
-				_ = op.UpdateMetadata(metadata)
+		revert.Success()
+		return nil
+	}
 
-				// Get the shutdown timeout for the instance.
-				timeout := inst.ExpandedConfig()["boot.host_shutdown_timeout"]
-				val, err := strconv.Atoi(timeout)
-				if err != nil {
-					val = evacuateHostShutdownDefaultTimeout
-				}
+	op, err := operations.OperationCreate(s, "", operations.OperationClassTask, operationtype.ClusterMemberEvacuate, nil, nil, run, nil, nil, r)
+	if err != nil {
+		return response.InternalError(err)
+	}
 
-				// Start with a clean shutdown.
-				err = inst.Shutdown(time.Duration(val) * time.Second)
-				if err != nil {
-					l.Warn("Failed shutting down instance, forcing stop", logger.Ctx{"err": err})
+	return operations.OperationResponse(op)
+}
 
-					// Fallback to forced stop.
-					err = inst.Stop(false)
-					if err != nil && !errors.Is(err, instanceDrivers.ErrInstanceIsStopped) {
-						return fmt.Errorf("Failed to stop instance %q in project %q: %w", inst.Name(), instProject.Name, err)
-					}
-				}
+func evacuateInstances(ctx context.Context, d *Daemon, r *http.Request, instances []instance.Instance, mode string, nodeName string, stopInstance evacuateStopFunc, migrateInstance evacuateMigrateFunc, op *operations.Operation) error {
```

Comment:
This argument list is getting a little long, how about introducing an `opts` struct. Also I dont think this needs *Daemon and can just use *State. Keep the ctx as an argument thought so `evacuateInstances(ctx, opts)`